### ISSUE TITLE
Is oneOf with $refs supported (see attached example)

### ISSUE THREAD

**Issue by SteveAtSentosa** [tdegrunt/jsonschema#157]
Thanks for a great tool.

I am wanting to create a Schema physDevice which can be one of a Set Tob Box (stb) or Cable Modem (CM) ... yes I work in the cable TV industry.

For the attached file these last two validations fail

``` json
  console.log("\n----- physDev oneOf w stb -----");
  console.log(v.validate(stbPL, physDevSchema));  

  console.log("\n----- physDev oneOf w CM -----");
  console.log(v.validate(cmPL, physDevSchema));  
```

I was expecting these to pass.  I was wondering if your tool supports oneOf in combinations with $ref? ... or do I just have the schema incorrectly defined?

[val-oneof-ref.js.txt](https://github.com/tdegrunt/jsonschema/files/111454/val-oneof-ref.js.txt)

**Comment by PauloASilva** [comment]
Hi,
I was having the same issue with `oneOf` but with a little difference: I was testing

```javascript
myProperty : {
    type: {
        oneOf: [
            {type: 'null'},
            {$ref: '/subschema'}
        ]
    }
}
```

Then I realize that, there was no reason to do not test it as I actually do when I want a `null` or `something`. So

```javascript
myProperty : {
    type: [ 'null', {$ref: '/subschema'}]
}
```
and it did work.

I did test also your example with few little changes

```javascript
var Validator = require('jsonschema').Validator;
var v = new Validator();

// to make sure simple ref to STB schema works
var stbRefSchema = {
"$schema": "http://json-schema.org/draft-04/schema#",
    "id": "/stbRef",
    "type": "object",
    "properties": {
      "device": { "$ref": "/stb"
      }
    }
  };

// to make sure simple ref to CM schema works
var cmRefSchema = {
"$schema": "http://json-schema.org/draft-04/schema#",
    "id": "/cableModemRef",
    "type": "object",
    "properties": {
      "device": { "$ref": "/cableModem"
      }
    }
  };

// one-of two possible refs
var physDevSchema = {
"$schema": "http://json-schema.org/draft-04/schema#",
    "id": "/physDev",
    "type": "object",
    "properties": {
      "device": {
        "type": [
          { "$ref": "/cableModem" },
          { "$ref": "/stb" }
        ]
      }
    }
  };

var stbSchema = {
  "$schema": "http://json-schema.org/draft-04/schema#",
    "id": "/stb",
    "type": "object",
    "properties": {
      "stbDeviceDescription": {
        "type": "string"
      }
    }
  };

var cmSchema = {
  "$schema": "http://json-schema.org/draft-04/schema#",
    "id": "/cableModem",
    "type": "object",
    "properties": {
      "cableModemDescription": {
        "type": "string"
      }
    }
  };

  var stbPL = {
    "device":  {
          "stbDeviceDescription": "whole home DVR"
    }
  };

  var cmPL = {
    "device": {
        "cableModemDescription": "Doscsis 3.1 w WiF"
    }
  };

v.addSchema( cmSchema,'/cableModem');
v.addSchema( stbSchema,'/stb');
v.addSchema( stbRefSchema,'/stbRef');
v.addSchema( cmRefSchema,'/cableModem');
v.addSchema( physDevSchema,'/physDev');

console.log("\n----- cable modem -----");
v.validate(cmPL, cmSchema, {throwError: true});

console.log("\n----- stb -----");
v.validate(stbPL, stbSchema, {throwError: true});

console.log("\n----- cable modem ref  -----");
v.validate(cmPL, cmRefSchema, {throwError: true});

console.log("\n----- stb ref -----");
v.validate(stbPL, stbRefSchema, {throwError: true});

console.log("\n----- physDev oneOf w stb -----");
v.validate(stbPL, physDevSchema, {throwError: true});

console.log("\n----- physDev oneOf w CM -----");
v.validate(cmPL, physDevSchema, {throwError: true});
```

and it works, also.

What do you think?

**Comment by SteveAtSentosa** [comment]
Thanks Paulo, much appreciated !!!

   Steven Saunders
   Founder and Principal
   Sentosa Technology Consultants
   Cell  303.809.8043
   Email  steve@sentosatech.com <mailto:steve@sentosatech.com>
   LinkedIn: click here <https://www.linkedin.com/pub/steven-saunders/5/a35/783>
   Web  www.sentosatech.com <http://www.sentosatech.com/>

**Comment by PauloASilva** [comment]
@SteveAtSentosa You're welcome ;)

**Comment by Acubed** [comment]
@SteveAtSentosa, @PauloASilva, do you need exactly one of the alternative schemas, or at least one of them (i.e. the `anyOf` property)?

**Comment by PauloASilva** [comment]
@Acubed In my use case I want `type` to be `null` or, when defined, to be  _of type_ `subschema`. To answer your question, I would say "at least one of them".

```javascript
myProperty : {
    type: [ 'null', {$ref: '/subschema'}]
}
```

I did try the `oneOf` first, because I saw it on some StackOverflow thread.
Is the `anyOf` option available/implemented?

**Comment by Acubed** [comment]
@SteveAtSentosa `anyOf` is implemented, and we claim to have full, 100% JSON Schema support.

I might look at adding `"required"` instead , unless you intend for the empty object `{}` to be a valid instance (or, do both).

**Comment by PauloASilva** [comment]
@SteveAtSentosa "unless you intend for the empty object {} to be a valid instance" I don't want that.

All my properties are `null` by default.
For a non required property `x`, it can be `null` (if not set) otherwise it has to be  _of type_ `subSchema`.

**Comment by ianbytchek** [comment]
Also getting this – works without `throwError`, doesn't when it's on. This is labelled with feedback, is this still the case, do you need more info?

**Comment by Acubed** [comment]
@ianbytchek `throwError` needs to be reworked some. What's your particular issue?

It was marked Feedback because I need to know if `anyOf` or adding `required` fixes the problem here, like it should.

When using `oneOf`, exactly one of the schemas must match, not zero and not two or more.

If you've got a question, feel free to open a new issue and post your regression with code example, and expected and actual output.

**Comment by ianbytchek** [comment]
@Acubed I was having a lot of trouble with nested `oneOf`, eventually I overrode the `validate` method with the following in TypeScript, by the way would be awesome to have the official `d.ts` file.

```ts
validate(instance:any, schema:any, options?:any, context?:any):Validation {
    var throwError:boolean = false;

    if (options != null && options.throwError) {
        throwError = true;
        options.throwError = false;
    }

    var result:Validation = super.validate(instance, schema, options, context);

    if (throwError && result.errors.length > 0) {
        throw Error(String(result.errors));
    }

    return result;
}
```

This must be the general solution. With stuff like `oneOf`, we can't tell if the data is valid or not until we check every single schema with `oneOf` rules, which is what happens with `{throwError: false}`.

The following schema will always fail with `{throwError: false}`. Changing outer `oneOf` to `anyOf` will make it work with any of the given types.

```json
{
    "id": "/tst",
    "oneOf": [
        {"$ref": "#/definitions/bar"},
        {"$ref": "#/definitions/foo"},
    ],
    "definitions": {
        "foo": {"type": "boolean"},
        "bar": {"oneOf": [{"type": ["string", "number"]}]}
    }
}
```

I haven't given it deeper thoughts and as much as I wish to provide a better contribution, I can't at the moment, the approach might be to create a context for every `oneOf` occurrence and pass validation stack depth into each child validation. At the top of the stack (the topmost `oneOf` occurrence) we check child validation results for errors – we must get `n - 1` failed items. Stack items with depth greater than zero cannot throw an error, only root can.

This is a little more optimised than the solution above. We don't have to check / modify configuration, plus by selectively targeting `oneOf` cases we don't have to wait for the entire data to validate to tell if there's a problem.

**Comment by ljharb** [comment]
I think this is the cause of my issue: I'm refactoring a top-level object schema to be a "oneOf" schema, so that i can support both an object, and an object containing an array of objects. However, the validation message ceases to be useful when I convert to `oneOf` (even with only the same single schema). Is this something that can be fixed? Should I file a new issue?

**Comment by ljharb** [comment]
oops, nvm, just discovered the `nestedErrors` option.

**Comment by awwright** [comment]
@SteveAtSentosa Did this issue get resolved for you?

**Comment by awwright** [comment]
I'm going to close this out for inactivity. Please, feel free to open a new issue if there's any further problems.
